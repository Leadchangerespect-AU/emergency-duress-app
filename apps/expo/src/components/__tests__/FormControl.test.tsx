import React from "react";
import { Text, View } from "react-native";
import { render, screen } from "@testing-library/react-native";

import {
  FormControl,
  FormControlError,
  FormControlErrorIcon,
  FormControlErrorText,
  FormControlLabel,
  useFormControlContext,
} from "../FormControl";

jest.mock("@expo/vector-icons", () => ({
  Ionicons: "",
}));

describe("FormControl", () => {
  it("renders children correctly", () => {
    render(
      <FormControl>
        <View testID="test-child" />
      </FormControl>,
    );
    expect(screen.getByTestId("test-child")).toBeTruthy();
  });

  it("applies custom className", () => {
    render(
      <FormControl className="custom-class">
        <View />
      </FormControl>,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("provides context values to children", () => {
    const TestChild = () => {
      const context = useFormControlContext();
      return (
        <View testID="test-child">
          {context.isInvalid && <Text>Invalid</Text>}
          {context.isDisabled && <Text>Disabled</Text>}
          {context.isRequired && <Text>Required</Text>}
        </View>
      );
    };

    render(
      <FormControl isInvalid isDisabled isRequired>
        <TestChild />
      </FormControl>,
    );

    expect(screen.getByText("Invalid")).toBeTruthy();
    expect(screen.getByText("Disabled")).toBeTruthy();
    expect(screen.getByText("Required")).toBeTruthy();
  });

  it("matches snapshot", () => {
    render(
      <FormControl>
        <View></View>
      </FormControl>,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });
});

describe("FormControlError", () => {
  it("renders error message when form is invalid", () => {
    render(
      <FormControl isInvalid>
        <FormControlError>
          <Text>Error message</Text>
        </FormControlError>
      </FormControl>,
    );
    expect(screen.getByText("Error message")).toBeTruthy();
  });

  it("does not render when form is valid", () => {
    render(
      <FormControl>
        <FormControlError>
          <Text>Error message</Text>
        </FormControlError>
      </FormControl>,
    );
    expect(screen.queryByText("Error message")).toBeNull();
  });

  it("does not render when form is disabled", () => {
    render(
      <FormControl isInvalid isDisabled>
        <FormControlError>
          <Text>Error message</Text>
        </FormControlError>
      </FormControl>,
    );
    expect(screen.queryByText("Error message")).toBeNull();
  });

  it("matches snapshot", () => {
    render(
      <FormControl isInvalid>
        <FormControlError>
          <Text>Error message</Text>
        </FormControlError>
      </FormControl>,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });
});

describe("FormControlErrorIcon", () => {
  it("renders with default props", () => {
    render(<FormControlErrorIcon />);
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("renders with custom props", () => {
    render(
      <FormControlErrorIcon
        name="warning-outline"
        size={32}
        color="#ff0000"
        className="custom-class"
      />,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });
});

describe("FormControlErrorText", () => {
  it("renders error text with default styles", () => {
    render(<FormControlErrorText>Error message</FormControlErrorText>);
    expect(screen.getByText("Error message")).toBeTruthy();
  });

  it("applies custom className", () => {
    render(
      <FormControlErrorText className="custom-class">
        Error message
      </FormControlErrorText>,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });
});

describe("FormControlLabel", () => {
  it("renders label text", () => {
    render(
      <FormControl>
        <FormControlLabel>Label Text</FormControlLabel>
      </FormControl>,
    );
    expect(screen.getByText("Label Text")).toBeTruthy();
  });

  it("shows required indicator when isRequired is true", () => {
    render(
      <FormControl isRequired>
        <FormControlLabel>Label Text</FormControlLabel>
      </FormControl>,
    );
    expect(screen.getByText("*")).toBeTruthy();
  });

  it("renders correctly", () => {
    render(<FormControlLabel>Web Label</FormControlLabel>);
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("applies custom className", () => {
    render(
      <FormControlLabel className="custom-class">Label Text</FormControlLabel>,
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });
});

describe("FormControl Integration Tests", () => {
  it("renders complete error state with icon and message", () => {
    render(
      <FormControl isInvalid>
        <FormControlError>
          <FormControlErrorIcon />
          <FormControlErrorText>Invalid input</FormControlErrorText>
        </FormControlError>
      </FormControl>,
    );
    expect(screen.getByText("Invalid input")).toBeTruthy();
  });

  it("does not render when form is valid", () => {
    render(
      <FormControl>
        <FormControlError>
          <FormControlErrorText>Error Content</FormControlErrorText>
        </FormControlError>
      </FormControl>,
    );
    expect(screen.queryByText("Error Content")).toBeNull();
  });

  it("does not render when form is disabled", () => {
    render(
      <FormControl isInvalid isDisabled>
        <FormControlError>
          <FormControlErrorText>Error Content</FormControlErrorText>
        </FormControlError>
      </FormControl>,
    );
    expect(screen.queryByText("Error Content")).toBeNull();
  });
});
