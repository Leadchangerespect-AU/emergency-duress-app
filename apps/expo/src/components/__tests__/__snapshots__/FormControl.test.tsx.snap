// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormControl applies custom className 1`] = `
<View
  className="my-2 custom-class"
>
  <View />
</View>
`;

exports[`FormControl matches snapshot 1`] = `
<View
  className="my-2"
>
  <View />
</View>
`;

exports[`FormControlError matches snapshot 1`] = `
<View
  className="my-2"
>
  <View
    className="flex flex-row items-center mt-2 gap-2"
  >
    <Text>
      Error message
    </Text>
  </View>
</View>
`;

exports[`FormControlErrorIcon renders with custom props 1`] = `
<
  className="custom-class"
  color="#ff0000"
  name="warning-outline"
  size={32}
/>
`;

exports[`FormControlErrorIcon renders with default props 1`] = `
<
  color="#ef4444"
  name="alert-circle-outline"
  size={24}
/>
`;

exports[`FormControlErrorText applies custom className 1`] = `
<Text
  className="text-red-500 text-sm custom-class"
>
  Error message
</Text>
`;

exports[`FormControlLabel applies custom className 1`] = `
<Text
  className="text-base text-neutral-800 font-semibold custom-class"
>
  Label Text
</Text>
`;

exports[`FormControlLabel renders correctly 1`] = `
<Text
  className="text-base text-neutral-800 font-semibold"
>
  Web Label
</Text>
`;
